### PR TITLE
grpc/transcoding: Add support for google.api.http_body response streaming

### LINKED ISSUE #5540: Support streaming HttpBody proto?
The support for handling google.api.HttpBody is cool, especially when I want to directly reply http response with audio `content_type`. I notice that streaming case is remained to be done, so may I ask you when will that be supported? It's really use for me in implementing speech synthesis because l can transfer large audio data to client by clipping them into audio clips with chunked audio/mp3 `content_type`.

### PR DESCRIPTION
**Pull request by belyalov**
Description: Add support for `httpBody` for response streaming use case.
Risk Level: Low
Testing: New unit / integration test
Docs Changes: Added
Release Notes: Added
Fixes #5540 

This is my first contribution to Envoy project, so I need your help to ramp up with code base.
Please check inline questions and help resolve them.

### CODE REVIEW

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc [hunk 1]
```diff
@@ -504,11 +503,7 @@ Http::FilterHeadersStatus JsonTranscoderFilter::encodeHeaders(Http::ResponseHead
   }
 
   headers.setReferenceContentType(Http::Headers::get().ContentTypeValues.Json);
-  if (!method_->descriptor_->server_streaming()) {
-    return Http::FilterHeadersStatus::StopIteration;
-  }
-
-  return Http::FilterHeadersStatus::Continue;
+  return Http::FilterHeadersStatus::StopIteration;
```
Comment (belyalov) [inline source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc:506]: I'm not sure that this is correct way of solving problem:

When filter returns with `Http::FilterHeadersStatus::Continue` it automatically adds(preserves?) `Content-Type: application/grpc` header.

I wasn't able to find correct(?) way of doing this, could you please confirm / advice a better way?

File: docs/root/configuration/http/http_filters/grpc_json_transcoder_filter.rst [hunk 2]
```diff
@@ -82,7 +82,10 @@ as its output message type. The implementation needs to set
 (which sets the value of the HTTP response `Content-Type` header) and
 `data <https://github.com/googleapis/googleapis/blob/master/google/api/httpbody.proto#L71>`_
 (which sets the HTTP response body) accordingly.
-
+Please note that when gRPC response streaming is used HTTP response header `Content-Type` will be set
```
Comment (qiwzhang) [inline docs/root/configuration/http/http_filters/grpc_json_transcoder_filter.rst:85]: Multiple Httpbody can be send by the gRPC server in the server streaming case. In this case, HTTP response header `Content-Type` will use the content-type from the first Httpbody.

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc (hunk 1 above)
Comment (qiwzhang) [reply source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc:506]: I think returning "Stop" is be fine. It needs to wait for encodeData() call to set the header content-type.   If continue, it will be too late for encodeData() to modify the headers

File: test/extensions/filters/http/grpc_json_transcoder/grpc_json_transcoder_integration_test.cc [hunk 3]
```diff
@@ -328,6 +328,22 @@ TEST_P(GrpcJsonTranscoderIntegrationTest, UnaryGetHttpBody) {
...
+          {":method", "GET"}, {":path", "/indexStream"}, {":authority", "host"}},
+      "", {""},
+      {R"(content_type: "text/html" data: "<h1>Hello!</h1>")",
+       R"(content_type: "text/plain" data: "Hello!")"},
+      Status(),
+      Http::TestResponseHeaderMapImpl{{":status", "200"}, {"content-type", "text/html"}},
+      R"(<h1>Hello!</h1>)"
+      R"(Hello!)"
+  );
+}
```
[the remark this answers is not in the document]
Comment (qiwzhang) [reply test/extensions/filters/http/grpc_json_transcoder/grpc_json_transcoder_integration_test.cc]: transcoder_ is adding [].  You should not getting any data from transcoder.

[the remark this answers is not in the document]
Comment (lizan) [reply test/extensions/filters/http/grpc_json_transcoder/grpc_json_transcoder_integration_test.cc]: specifically, https://github.com/envoyproxy/envoy/blob/master/source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc#L559

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc (hunk 1 above)
Comment (belyalov) [reply source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc]: Thanks for feedback!
Updated accordingly.

File: test/extensions/filters/http/grpc_json_transcoder/grpc_json_transcoder_integration_test.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (belyalov) [reply test/extensions/filters/http/grpc_json_transcoder/grpc_json_transcoder_integration_test.cc]: Now fixed.
Thanks for help!

File: docs/root/configuration/http/http_filters/grpc_json_transcoder_filter.rst (hunk 2 above)
Comment (belyalov) [reply docs/root/configuration/http/http_filters/grpc_json_transcoder_filter.rst]: Rephrased.

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc [hunk 4]
```diff
@@ -504,7 +503,9 @@ Http::FilterHeadersStatus JsonTranscoderFilter::encodeHeaders(Http::ResponseHead
   }
 
   headers.setReferenceContentType(Http::Headers::get().ContentTypeValues.Json);
-  if (!method_->descriptor_->server_streaming()) {
+
+  if (!method_->descriptor_->server_streaming() ||
+      (method_->descriptor_->server_streaming() && method_->response_type_is_http_body_)) {
```
Comment (qiwzhang) [inline source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc:508]: logically, you don't need the 2nd server_streaming check

Comment (belyalov) [reply source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc]: Thanks for catch, simplified.

**Comment by belyalov** [comment]
BTW, what is the preferred way to resolve conflicts?
Is it OK to rebase once you've done with review or merge is a way to go?

**Comment by euroelessar** [comment]
> BTW, what is the preferred way to resolve conflicts?
> Is it OK to rebase once you've done with review or merge is a way to go?

Please check https://github.com/envoyproxy/envoy/blob/master/CONTRIBUTING.md#submitting-a-pr.
TL;DR Please only merge, do not rebase.

**Comment by belyalov** [comment]
All done! :)

**Comment by belyalov** [comment]
@lizan please let me know if something else needs to be done to get this merged.. )

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc [hunk 5]
```diff
@@ -502,11 +501,11 @@ Http::FilterHeadersStatus JsonTranscoderFilter::encodeHeaders(Http::ResponseHead
   }
 
   headers.setReferenceContentType(Http::Headers::get().ContentTypeValues.Json);
-  if (!method_->descriptor_->server_streaming()) {
-    return Http::FilterHeadersStatus::StopIteration;
-  }
 
-  return Http::FilterHeadersStatus::Continue;
+  if (method_->descriptor_->server_streaming() && !method_->response_type_is_http_body_) {
```
Comment (lizan) [inline source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc:505]: Worth a comment why this condition.

Comment (belyalov) [reply source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc:508]: Agree, explained.

**Comment by belyalov** [comment]
@lizan ping?

**Comment by lizan** [comment]
Sorry we need another master merge to fix conflict in version history. Thanks

**Comment by belyalov** [comment]
> Sorry we need another master merge to fix conflict in version history. Thanks

Resolved! :)

**Comment by npuichigo** [comment]
it seems that coverage test failed.

**Comment by mattklein123** [comment]
Please merge master (this should also fix coverage).

/wait

**Comment by belyalov** [comment]
> Please merge master (this should also fix coverage).
> 
> /wait

Sure, merged and fixed
All checks are now done!

**Comment by npuichigo** [comment]
@lizan ping

**Comment by npuichigo** [comment]
@belyalov I tested this pr to transcode a streaming grpc to chunked http response, but the content-type of curl is not correct.
```shell
curl "localhost:8080/get?filename=testdata/music.mp3" -4 -v > test_chunk_music.mp3
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 8080 (#0)
> GET /get?filename=testdata/music.mp3 HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.47.0
> Accept: */*
>
< HTTP/1.1 200 OK
< content-type: application/json
< grpc-accept-encoding: identity,deflate,gzip
< accept-encoding: identity,gzip
< x-envoy-upstream-service-time: 0
< date: Thu, 23 Apr 2020 14:15:00 GMT
< server: envoy
< transfer-encoding: chunked
<
{ [14238 bytes data]
100 1999k    0 1999k    0     0   640k      0 --:--:--  0:00:03 --:--:--  639k^C
``` 
If I only send the chunk of data once, the content-type is correct.
```shell
% Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 8080 (#0)
> GET /get?filename=testdata/musc.mp3 HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.47.0
> Accept: */*
>
< HTTP/1.1 200 OK
< content-type: audio/mp3
< grpc-accept-encoding: identity,deflate,gzip
< accept-encoding: identity,gzip
< x-envoy-upstream-service-time: 0
< date: Thu, 23 Apr 2020 14:18:19 GMT
< server: envoy
< transfer-encoding: chunked
```
Is that the subsequent response chunk overwrite the content-type set by the first chunk?

**Comment by qiwzhang** [comment]
hmm,  this [line](https://github.com/envoyproxy/envoy/blob/4283e25cacb89627819be51b1c1cd7d4bfcb35ff/source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc#L511) 
should return StopIteration.

You can try it.

**Comment by belyalov** [comment]
> hmm, this [line](https://github.com/envoyproxy/envoy/blob/4283e25cacb89627819be51b1c1cd7d4bfcb35ff/source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc#L511)
> should return StopIteration.
> 
> You can try it.

Hm, it does return:
https://github.com/envoyproxy/envoy/blob/62777e87edc9f16af5c7920bb86f198585789a3d/source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc#L503-L511

**Comment by qiwzhang** [comment]
@npuichigo  could you debug it with "-l debug" log?

**Comment by belyalov** [comment]
@npuichigo could you please verify that you're using latest version from master?

I've tried to reproduce it on my test setup, it seems works well:

Having golang as gRPC server
```go
func (k *KV) GetHttp(in *kv.GetRequest, stream kv.KV_GetHttpServer) error {
	log.Printf("get: %s", in.Key)

	body := &httpbody.HttpBody{
		ContentType: "text/plain",
		Data:        []byte("data\n"),
	}
	stream.Send(body)
	body.ContentType = "unknown"
	stream.Send(body)

	return nil
}
```

Envoy configured with default gRPC transcoding:
```yaml
#...
          - name: envoy.filters.http.grpc_json_transcoder
            typed_config:
              "@type": type.googleapis.com/envoy.config.filter.http.transcoder.v2.GrpcJsonTranscoder
              proto_descriptor: "kv.pb"
              services: ["kv.KV"]
              print_options:
                add_whitespace: true
                always_print_primitive_fields: true
                always_print_enums_as_ints: false
                preserve_proto_field_names: false
#...
```

Request / Response:
```bash
$ curl -v http://localhost:8080/getHttp
*   Trying ::1...
* TCP_NODELAY set
* Connection failed
* connect to ::1 port 8080 failed: Connection refused
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 8080 (#0)
> GET /getHttp HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.64.1
> Accept: */*
>
< HTTP/1.1 200 OK
< content-type: text/plain
< x-envoy-upstream-service-time: 1
< date: Fri, 24 Apr 2020 01:17:11 GMT
< server: envoy
< transfer-encoding: chunked
<
data
data
* Connection #0 to host localhost left intact
* Closing connection 0
```

**Comment by npuichigo** [comment]
@belyalov I use the latest docker image of envoy-dev. I will verify that again.

**Comment by npuichigo** [comment]
@belyalov I put an example here to reproduce it (https://github.com/npuichigo/grpc_with_envoy). Here I re-pull the latest docker image of envoy-dev.

I tried to set content-type once and it works well. Multiple chunked responses even with the same content-type make the received content-type application/json, but I checked the received data and it's right without wrapping by json.

```bash
$ curl "localhost:8080/get?filename=testdata/small.mp3" -4 -v > test_chunk_music.mp3
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 8080 (#0)
> GET /get?filename=testdata/small.mp3 HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.47.0
> Accept: */*
>
< HTTP/1.1 200 OK
< content-type: audio/mp3
< grpc-accept-encoding: identity,deflate,gzip
< accept-encoding: identity,gzip
< x-envoy-upstream-service-time: 0
< date: Fri, 24 Apr 2020 03:26:57 GMT
< server: envoy
< transfer-encoding: chunked
<

$ curl "localhost:8080/get?filename=testdata/music.mp3" -4 -v > test_chunk_music.mp3
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 8080 (#0)
> GET /get?filename=testdata/music.mp3 HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.47.0
> Accept: */*
>
< HTTP/1.1 200 OK
< content-type: application/json
< grpc-accept-encoding: identity,deflate,gzip
< accept-encoding: identity,gzip
< x-envoy-upstream-service-time: 0
< date: Fri, 24 Apr 2020 03:27:24 GMT
< server: envoy
< transfer-encoding: chunked
<
```

**Comment by npuichigo** [comment]
@belyalov I put an example here to reproduce it (https://github.com/npuichigo/grpc_with_envoy). Here I re-pull the latest docker image of envoy-dev.

I found that it's cause by the size of HttpBody response. If the body size is smaller than 16KB, then it works well. Otherwise, the received content type is application/json. So is it by design and is there any setting to configure it?

You can change `static const int kBufferSize = 64 * 1024;` in https://github.com/npuichigo/grpc_with_envoy/blob/master/grpc_with_envoy/serving/demo_service_impl.cc to reproduce that.

**Comment by qiwzhang** [comment]
I see the problem is in this line

https://github.com/envoyproxy/envoy/blob/master/source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc#L526

If buildResponseFromHttpBodyOutput() did not receive a full HttpBody data frame, it should not return "continue",  it should return Stop to keep buffering the data.

**Comment by npuichigo** [comment]
@belyalov

**Comment by npuichigo** [comment]
@qiwzhang So do you think buildResponsFromHttpBodyOutput should return a boolean to indicate whether we receive a full data frame? Or do we need to check that only for HttpBody?
```
if (frames.empty()) {
    return true;
}
```

**Comment by belyalov** [comment]
@npuichigo and @qiwzhang 
Sorry guys, I missed this thread.

Got the repro and @qiannawang is right - basically what happens where is when `Http::FilterHeadersStatus::Continue` returned - envoy sends headers, that's why you see `application/json` as content type.

Going to fix it very soon, thanks for research!
